python/pfs/utils/optimize_black_dots.py

```diff
@@ -0,0 +1,482 @@
+                list_of_actual_position_x_observed.append(np.nan)
+                list_of_actual_position_y_observed.append(np.nan)
+
+                list_of_predicted_position_x_not_observed.append(np.nan)
+                list_of_predicted_position_y_not_observed.append(np.nan)
+
+        prepared_observations_and_predictions = [list_of_actual_position_x_observed,
+                                                 list_of_actual_position_y_observed,
+                                                 list_of_predicted_position_x_not_observed,
+                                                 list_of_predicted_position_y_not_observed]
+
+        return prepared_observations_and_predictions
+
+    def rotate(self, point, origin, degrees):
+        """rotate the point around the origin
+
+        gets called by `new_position_of_dot`
+
+        Parameters
+        ----------
+        point: `tuple`
+            x, y coordinate of the point to rotate
+        origin: `tuple`
+            x, y coordinate of the point around which to rotate
+        degrees: `float`
+            number of degrees to rotate
+
+        Returns
+        ----------
+        x, y coordinate of the rotated point
+        """
+        radians = np.deg2rad(degrees)
+        x, y = point
+        offset_x, offset_y = origin
+        adjusted_x = (x - offset_x)
+        adjusted_y = (y - offset_y)
+        cos_rad = np.cos(radians)
+        sin_rad = np.sin(radians)
+        qx = offset_x + cos_rad * adjusted_x + sin_rad * adjusted_y
+        qy = offset_y + -sin_rad * adjusted_x + cos_rad * adjusted_y
+        return qx, qy
+
+    def check(self, x, y, xc, yc, r=None):
+        """check if the cobra is covered by the black dot
+
+        gets called by `quality_measure_single_spot`
+
+        Parameters
+        ----------
+        x: `float`
+            x coordinates of the dot
+        y: `float`
+            y coordinate of the dot
+        xc: `float`
+            x coordinate of the cobra
+        yc: `float`
+            y coordinate of the cobra
+        r: `float`
+            radius of dot [mm]
+
+        Returns
+        ----------
+        `int`
+         returns 1 if the cobra is covered, 0 if not
+        """
+        if r is None:
+            r = 0.75
+        return int(((x - xc)**2 + (y - yc)**2) < r**2)
+
+    def new_position_of_dot(self, i, xd_mod, yd_mod, scale, rot, x_scale_rot, y_scale_rot):
+        """move the dot to the new position
+
+        This function defines the change of the positions that we consider
+
+        calls `rotate`
+
+        Parameters
+        ----------
+        i: `integer`
+            index of the cobra
+        xd_mod: `float`
+            x offset of the new position
+        yd_mod: `float`
+            y offset of the new position
+        scale: `float`
+            overall scaling
+        x_scale_rot: `float`
+            x coordinate around which to rotate
```

Comment:
The x-coordinate is scaled, though, correct? If not, why is the word 'scale' used in the variable?